lib/scholar/decomposition/pca.ex

```diff
@@ -0,0 +1,216 @@
+      but can sometime improve the predictive accuracy of the downstream estimators by making their data respect some hard-wired assumptions.
+
+      """
+    ]
+  ]
+
+  @fit_opts_schema NimbleOptions.new!(fit_opts_schema)
+  @transform_opts_schema NimbleOptions.new!(transform_opts_schema)
+
+  @doc """
+  Fits a PCA for sample inputs `x`.
+
+  ## Options
+
+  #{NimbleOptions.docs(@fit_opts_schema)}
+
+  ## Returns
+
+  The function returns a struct with the following parameters:
+
+    * `:components` - Principal axes in feature space, representing the directions of maximum variance in the data.
+      Equivalently, the right singular vectors of the centered input data, parallel to its eigenvectors.
+      The components are sorted by `:explained_variance`.
+
+    * `:explained_variance` - The amount of variance explained by each of the selected components.
+      The variance estimation uses `:num_samples - 1` degrees of freedom.
+      Equal to `:num_components` largest eigenvalues of the covariance matrix of `x`.
+
+    * `:explained_variance_ratio` - Percentage of variance explained by each of the selected components.
+      If `:num_components` is not set then all components are stored and the sum of the ratios is equal to 1.0.
+
+    * `:singular_values` - The singular values corresponding to each of the selected components.
+      The singular values are equal to the 2-norms of the `:num_components` variables in the lower-dimensional space.
+
+    * `:mean` - Per-feature empirical mean, estimated from the training set.
+
+    * `:num_components` - The estimated number of components. When `:num_components` is set to a number between 0 and 1
+      this number is estimated from input data. Otherwise it equals the parameter `:num_components`, or the lesser
+      value of `:num_features` and `:num_samples` if `num_components` is `nil`.
+
+    * `:num_features` - Number of features in the training data.
+
+    * `:num_samples` - Number of samples in the training data.
+
+    * `:decomposer` - Matrix used in actual decomposition. It is the unitary matrix from SVD factorization.
+  """
+  deftransform fit(x, opts \\ []) do
+    fit_n(x, NimbleOptions.validate!(opts, @fit_opts_schema))
+  end
+
+  defnp fit_n(x, opts \\ []) do
+    if Nx.rank(x) != 2 do
+      raise ArgumentError, "expected x to have rank equal to: 2, got: #{inspect(Nx.rank(x))}"
+    end
+
+    {num_samples, num_features} = Nx.shape(x)
+    num_components = opts[:num_components]
+
+    mean = Nx.mean(x, axes: [0])
+    x = x - mean
+    {decomposer, singular_values, components} = Nx.LinAlg.svd(x)
+    explained_variance = singular_values * singular_values / (num_samples - 1)
+    explained_variance_ratio = explained_variance / Nx.sum(explained_variance)
+
+    num_components =
+      calculate_num_components(
+        num_components,
+        num_features,
+        num_samples,
+        explained_variance
+      )
```

Comment:
I think it's confusing if the returned values include all components, but `num_components` is smaller. Even more so, that after `fit` the user needs to do `PCA.transform(num_components: Nx.to_number(model.num_components))`.

I think we can either:

1. Only accept `:num_components` as an integer and slice all the values accordingly (and revisit once/if we have dynamic shapes?)
2. Have a separate function that takes the desired explained variance and reduces the number of components accordingly. This would need to be non-defn because the shape is dynamic

Opinions? cc @polvalente @seanmor5

Reply:
Should we then change transform to get the num_components from the model instead?

Reply:
@josevalim we would need to call `Nx.to_number` inside that transform, so it would no longer make sense as a transform.

Reply:
So we need to compute `num_components` as a `defn`?

Reply:
If we compute num_components as defn, then num_components will be a tensor, and the problem still occurs.

Reply:
@josevalim the thing is that in sklearn `num_components` can be either a static number or a 0.0-1.0 float, in which case they compute `num_components` dynamically to preserve the given amount of variance. So we need `num_components` to be static, but it is dynamic in the float case.

Reply:
I pushed a commit that makes `num_components` a dynamic value on `fit`. The idea was to force dynamic values so we force thinking about the dynamic nature of the problem.

For `transform`, my suggestion is to rename `transform` to `mask` and convert all non component values to zero. By masking the decomposed values, we leave the dim reduction to the caller who will hopefully have more knowledge about the shapes. Would that makes sense?

Reply:
Btw, if you want to keep my commit, then please make sure to add tests that both integer and float values for num_components are clipped.

Reply:
> By masking the decomposed values, we leave the dim reduction to the caller who will hopefully have more knowledge about the shapes. Would that makes sense?

I'm not sure which approach is best here. If a user has a dataset with 100 features and wants to reduce it to 10 features with PCA, then by masking we don't really reduce the dimensionality, so they may end up with unnecessarily heavy computation later on. They could slice the transformed dataset themselves, but it's kinda weird to say `num_components: 10` just to manually slice 10 afterwards.

I can't say much about the typical use case, but I would assume that the dynamic case is more useful during manual exploration, rather than a part of a bigger algorithm. So I would stick to static `num_components` and for exploration they can do this:

```elixir
model = Scholar.Decomposition.PCA.fit(data)
Nx.cumulative_sum(model.explained_variance)
# See which point is best to drop components
model = Scholar.Decomposition.PCA.fit(data, num_components: 4)
```

We can also add a separate function for slicing the model if somebody requests it.

> For transform, my suggestion is to rename transform to mask and convert all non component values to zero.

I'd stick with `transform`, because even if we mask some features, the other ones change :)

Reply:
So we could do this:

* if `num_components` is integer or :none, we store it in the model as a static value
* if `float`, we don't store `num_components`
* On transform, you can pass your own or we fallback to `num_components`. If `num_components` is nil, we raise

Thoughts?

Alternatively we don't support `float` `num_components`at all and revert my changes.

Reply:
> if `float`, we don't store `num_components`

In that case I don't think we end up using the float at all.

I would go with static `num_components` (dropping support for floats) :)